mapper/orm/src/main/java/org/hibernate/search/mapper/orm/massindexing/impl/IdentifierConsumerDocumentProducer.java

```diff
@@ -196,39 +197,85 @@ private void indexAllQueue(SessionImplementor session, List<?> entities) throws
 			return;
 		}
 
-		// TODO use 'SessionWorkExecutor workExecutor = mapping.createSearchManager( session ).createSessionWorkExecutor();' instead
-		PojoWorkPlan workPlan = mapping.createSearchManager( session ).createWorkPlan();
+		// TODO <<monitor>> indexingMonitor.entitiesLoaded( entities.size() );
+		PojoSessionWorkExecutor workExecutor = mapping.createSearchManager( session ).createSessionWorkExecutor();
+		CompletableFuture<?>[] tasks = new CompletableFuture<?>[entities.size()];
 
-		log.tracef( "received a list of objects to index: %s", entities );
-		for ( Object object : entities ) {
-			try {
-				index( workPlan, object );
-				// TODO: implements monitor
-				// monitor.documentsBuilt( 1 );
-			}
-			catch (RuntimeException e) {
-				String errorMsg = log.massIndexerUnableToIndexInstance(
-						object.getClass().getName(),
-						object.toString()
-				);
-
-				// TODO: implements exception handler
-				// errorHandler.handleException( errorMsg, e );
-				// temporary re-throw the exception
-				throw new RuntimeException( errorMsg, e );
+		for ( int i = 0; i < entities.size(); i++ ) {
+			tasks[i] = index( workExecutor, entities.get( i ) );
+		}
+
+		try {
+			CompletableFuture.allOf( tasks ).join();
+		}
+		catch (RuntimeException e) {
+			Integer taskWithExceptionIndex = findAnyCompletedExceptionallyTask( tasks );
+			if ( taskWithExceptionIndex == null ) {
+				throw new RuntimeException( "This row should be never reached, according to CompletableFuture#allOf API", e );
 			}
+
+			// handling error on executing task phase
+			handleException( entities.get( taskWithExceptionIndex ), e );
 		}
 
-		workPlan.execute().get();
+		Integer completedTasks = this.countCompletedTasks( tasks );
```

Comment:
Good point. In Search 5 we used to increment the monitor even if a work failed, but we probably should not.